### DISCUSSION TITLE
Improve test coverage

### DISCUSSION THREAD

**Discussion by JaredWright** [Bareflank/pal#95] [Ideas]
As of now, testing in the PAL project is entirely focused on checking that the project's code generator produces code that will actually compile, and that generated functions follow naming conventions correctly. There are other aspects of this project that could potentially benefit from testing as well:

1) Test the implementation of instructions in the `pal/libpal` directory.
2) Test the format/fields of .yml files in the project's database

@zhaofengli threw out the idea of running PAL instructions in an emulator like bochs to test if they execute correctly and return expected values. This seems possible, but would have some challenges to think through first. For example, what kind of software execution environment would the tests run in? Would there need to be a minimal OS to support execution in a baremetal context on bochs? What would be the scope and goal of the tests (e.g. cover all the instructions in all libpal targets? just baremetal ones?)

I'd love to see more ideas in this discussion!

**Comment by zhaofengli** [comment]
> Would there need to be a minimal OS to support execution in a baremetal context on bochs?

With the things that we are testing (reading/setting various registers), a very simple kernel should suffice. For example, we have [this tiny kernel](https://github.com/mars-research/realjump/tree/b489a7cda40b39ce5f4e77a7ce7b754ca87b2b52/test-os) for a small kernel-level library that we are working on. It's built and run as part of the [CI pipeline](https://github.com/mars-research/realjump/blob/b489a7cda40b39ce5f4e77a7ce7b754ca87b2b52/.github/workflows/build.yml).
